handlers/common.go

```diff
@@ -130,27 +128,214 @@ func GetValidatorEarnings(validators []uint64, currency string) (*types.Validato
+	// Return 0 if there are no proposed blocks or no validators
+	if len(slots) == 0 || validatorsCount == 0 {
+		return 0
+	}
+	// Timeframe constants
+	fiveDays := time.Hour * 24 * 5
+	oneWeek := time.Hour * 24 * 7
+	oneMonth := time.Hour * 24 * 30
+	sixWeeks := time.Hour * 24 * 45
+	twoMonths := time.Hour * 24 * 60
+	threeMonths := time.Hour * 24 * 90
+	fourMonths := time.Hour * 24 * 120
+	fiveMonths := time.Hour * 24 * 150
+
+	activeValidatorsCount := *services.GetLatestStats().ActiveValidatorCount
+	// Calculate the expected number of slot proposals for 30 days
+	expectedSlotProposals := calcExpectedSlotProposals(oneMonth, validatorsCount, activeValidatorsCount)
+
+	// Get the timeframe for which we should consider qualified proposals
+	var proposalTimeframe time.Duration
+	// Time since the first block in the proposed block slice
+	timeSinceFirstBlock := time.Since(utils.SlotToTime(slots[0]))
+
+	// Determine the appropriate timeframe based on the time since the first block and the expected slot proposals
+	switch {
+	case timeSinceFirstBlock < fiveDays:
+		proposalTimeframe = fiveDays
+	case timeSinceFirstBlock < oneWeek:
+		proposalTimeframe = oneWeek
+	case timeSinceFirstBlock < oneMonth:
+		proposalTimeframe = oneMonth
+	case timeSinceFirstBlock > fiveMonths && expectedSlotProposals <= 0.75:
+		proposalTimeframe = fiveMonths
+	case timeSinceFirstBlock > fourMonths && expectedSlotProposals <= 1:
+		proposalTimeframe = fourMonths
+	case timeSinceFirstBlock > threeMonths && expectedSlotProposals <= 1.4:
+		proposalTimeframe = threeMonths
+	case timeSinceFirstBlock > twoMonths && expectedSlotProposals <= 2.1:
+		proposalTimeframe = twoMonths
+	case timeSinceFirstBlock > sixWeeks && expectedSlotProposals <= 2.8:
+		proposalTimeframe = sixWeeks
+	default:
+		proposalTimeframe = oneMonth
+	}
+
+	// Recalculate expected slot proposals for the new timeframe
+	expectedSlotProposals = calcExpectedSlotProposals(proposalTimeframe, validatorsCount, activeValidatorsCount)
+	if expectedSlotProposals == 0 {
+		return 0
+	}
+	// Cutoff time for proposals to be considered qualified
+	blockProposalCutoffTime := time.Now().Add(-proposalTimeframe)
+
+	// Count the number of qualified proposals
+	qualifiedProposalCount := 0
+	for _, slot := range slots {
+		if utils.SlotToTime(slot).After(blockProposalCutoffTime) {
+			qualifiedProposalCount++
+		}
+	}
+	// Return the luck as the ratio of qualified proposals to expected slot proposals
+	return float64(qualifiedProposalCount) / expectedSlotProposals
+}
+
+// calcExpectedSlotProposals calculates the expected number of slot proposals for a certain time frame and validator count
+func calcExpectedSlotProposals(timeframe time.Duration, validatorCount int, activeValidatorsCount uint64) float64 {
+	if validatorCount == 0 || activeValidatorsCount == 0 {
+		return 0
+	}
+	slotsPerTimeframe := timeframe.Seconds() / float64(utils.Config.Chain.Config.SecondsPerSlot)
```

Comment:
**Minor:** I think `slotsPerTimeframe` is a bad name since it's not "slots per time" but just a slot count, how about `slotsInTimeframe`?

Reply:
I honestly don't even know where the difference between "in" and "per" would be, but addressed in new commit, ptal